### ISSUE TITLE
daemon crashed after receiving HTTP 413 from registry

### ISSUE THREAD

**Issue by kdomanski** [docker/docker#9913]
If a private registry's maximum request size is reached (such as enforced by **client_max_body_size** in **nginx** config), the registry will abort and return HTTP code **413**. This crashes the docker engine.

I can reproduce on both **1.3.2** and **1.4.1**.

Log from 1.4.1:

```
INFO[0000] +job serveapi(unix:///var/run/docker.sock)   
INFO[0000] Listening for HTTP on unix (/var/run/docker.sock) 
INFO[0000] +job init_networkdriver()                    
INFO[0000] -job init_networkdriver() = OK (0)           
INFO[0000] Loading containers: start.                   
......................................................................................................................................................................
INFO[0000] Loading containers: done.                    
INFO[0000] docker daemon: 1.4.1 5bc2ff8; execdriver: native-0.2; graphdriver: btrfs 
INFO[0000] +job acceptconnections()                     
INFO[0000] -job acceptconnections() = OK (0)            
INFO[0001] POST /v1.16/images/[REDACTED]/[REDACTED]/[REDACTED]/push?tag= 
INFO[0001] +job push([REDACTED]/[REDACTED]/[REDACTED]) 
Received HTTP code 413 while uploading layer: <html>
<head><title>413 Request Entity Too Large</title></head>
<body bgcolor="white">
<center><h1>413 Request Entity Too Large</h1></center>
<hr><center>nginx/1.6.2</center>
</body>
</html>

INFO[0002] -job push([REDACTED]) = ERR (1) 
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x20 pc=0x70645c]

goroutine 95 [running]:
runtime.panic(0xbe5c40, 0x127da13)
        /usr/local/go/src/pkg/runtime/panic.c:279 +0xf5
bufio.(*Writer).flush(0xc2082c4e80, 0x0, 0x0)
        /usr/local/go/src/pkg/bufio/bufio.go:530 +0xdc
bufio.(*Writer).Flush(0xc2082c4e80, 0x0, 0x0)
        /usr/local/go/src/pkg/bufio/bufio.go:519 +0x39
net/http.(*response).Flush(0xc2081040a0)
        /usr/local/go/src/pkg/net/http/server.go:1023 +0x4c
github.com/docker/docker/utils.(*WriteFlusher).Write(0xc2082e87b0, 0xc20830cb40, 0xcc, 0x135, 0xcc, 0x0, 0x0)
        /go/src/github.com/docker/docker/utils/utils.go:265 +0x142
github.com/docker/docker/engine.(*Output).Write(0xc2082c5240, 0xc20830cb40, 0xcc, 0x135, 0x0, 0x0, 0x0)
        /go/src/github.com/docker/docker/engine/streams.go:90 +0x183
github.com/docker/docker/utils.(*WriteFlusher).Write(0xc208435e00, 0xc20830cb40, 0xcc, 0x135, 0x0, 0x0, 0x0)
        /go/src/github.com/docker/docker/utils/utils.go:264 +0xf4
github.com/docker/docker/utils.(*WriteFlusher).Write(0xc2083dbef0, 0xc20830cb40, 0xcc, 0x135, 0x0, 0x0, 0x0)
        /go/src/github.com/docker/docker/utils/utils.go:264 +0xf4
github.com/docker/docker/utils.(*WriteFlusher).Write(0xc208ae7260, 0xc20830cb40, 0xcc, 0x135, 0x0, 0x0, 0x0)
        /go/src/github.com/docker/docker/utils/utils.go:264 +0xf4
github.com/docker/docker/utils.(*progressReader).Read(0xc20879af00, 0xc208e7e000, 0x8000, 0x8000, 0x4218d5, 0x7fbc65d0b010, 0xc208ca83f0)
        /go/src/github.com/docker/docker/utils/progressreader.go:31 +0x2a9
github.com/docker/docker/vendor/src/code.google.com/p/go/src/pkg/archive/tar.(*regFileReader).Read(0xc208e2d7e0, 0xc208e7e000, 0x8000, 0x8000, 0xc208341200, 0x0, 0x0)
        /go/src/github.com/docker/docker/vendor/src/code.google.com/p/go/src/pkg/archive/tar/reader.go:748 +0xd8
github.com/docker/docker/vendor/src/code.google.com/p/go/src/pkg/archive/tar.(*Reader).Read(0xc20840c480, 0xc208e7e000, 0x8000, 0x8000, 0x7fbc64248840, 0x0, 0x0)
        /go/src/github.com/docker/docker/vendor/src/code.google.com/p/go/src/pkg/archive/tar/reader.go:733 +0xa4
github.com/docker/docker/pkg/tarsum.(*tarSum).Read(0xc208823d40, 0xc208e8a000, 0x8000, 0x8000, 0x5dfd1f, 0x0, 0x0)
        /go/src/github.com/docker/docker/pkg/tarsum/tarsum.go:152 +0x162
io.(*teeReader).Read(0xc208463580, 0xc208e8a000, 0x8000, 0x8000, 0xaac3c0, 0x0, 0x0)
        /usr/local/go/src/pkg/io/io.go:486 +0x8d
io/ioutil.(*nopCloser).Read(0xc208457540, 0xc208e8a000, 0x8000, 0x8000, 0x7b51, 0x0, 0x0)
        <autogenerated>:4 +0x7b
io.Copy(0x7fbc64218908, 0xc2083fece0, 0x7fbc65d1c618, 0xc208457540, 0x35daed, 0x0, 0x0)
        /usr/local/go/src/pkg/io/io.go:353 +0x1f3
net/http.(*transferWriter).WriteBody(0xc20871ce00, 0x7fbc65d1c0a0, 0xc208877040, 0x0, 0x0)
        /usr/local/go/src/pkg/net/http/transfer.go:202 +0x208
net/http.(*Request).write(0xc208823e10, 0x7fbc65d1c0a0, 0xc208877040, 0xc60300, 0xc208ae7e60, 0x0, 0x0)
        /usr/local/go/src/pkg/net/http/request.go:435 +0x90e
net/http.(*persistConn).writeLoop(0xc2085a3a20)
        /usr/local/go/src/pkg/net/http/transport.go:891 +0x214
created by net/http.(*Transport).dialConn
        /usr/local/go/src/pkg/net/http/transport.go:601 +0x957

goroutine 16 [chan receive]:
github.com/docker/docker/api/server.ServeApi(0xc20804ec80, 0x7fbc65d0a8e8)
        /go/src/github.com/docker/docker/api/server/server.go:1573 +0x473
github.com/docker/docker/engine.(*Job).Run(0xc20804ec80, 0x0, 0x0)
        /go/src/github.com/docker/docker/engine/job.go:83 +0x837
main.mainDaemon()
        /go/src/github.com/docker/docker/docker/daemon.go:91 +0x568
main.main()
        /go/src/github.com/docker/docker/docker/docker.go:71 +0x3ff

goroutine 19 [finalizer wait]:
runtime.park(0x41e470, 0x129d178, 0x1282de9)
        /usr/local/go/src/pkg/runtime/proc.c:1369 +0x89
runtime.parkunlock(0x129d178, 0x1282de9)
        /usr/local/go/src/pkg/runtime/proc.c:1385 +0x3b
runfinq()
        /usr/local/go/src/pkg/runtime/mgc0.c:2644 +0xcf
runtime.goexit()
        /usr/local/go/src/pkg/runtime/proc.c:1445

goroutine 20 [syscall]:
os/signal.loop()
        /usr/local/go/src/pkg/os/signal/signal_unix.go:21 +0x1e
created by os/signal.init·1
        /usr/local/go/src/pkg/os/signal/signal_unix.go:27 +0x32

goroutine 17 [syscall]:
runtime.goexit()
        /usr/local/go/src/pkg/runtime/proc.c:1445

goroutine 42 [chan receive]:
database/sql.(*DB).connectionOpener(0xc208188980)
        /usr/local/go/src/pkg/database/sql/sql.go:583 +0x48
created by database/sql.Open
        /usr/local/go/src/pkg/database/sql/sql.go:442 +0x27c

goroutine 23 [chan receive]:
github.com/docker/docker/pkg/signal.func·002()
        /go/src/github.com/docker/docker/pkg/signal/trap.go:30 +0x88
created by github.com/docker/docker/pkg/signal.Trap
        /go/src/github.com/docker/docker/pkg/signal/trap.go:53 +0x28f

goroutine 45 [sleep]:
time.Sleep(0x12a05f200)
        /usr/local/go/src/pkg/runtime/time.goc:39 +0x31
net.func·019()
        /usr/local/go/src/pkg/net/dnsclient_unix.go:183 +0x56
created by net.loadConfig
        /usr/local/go/src/pkg/net/dnsclient_unix.go:212 +0x153

goroutine 25 [IO wait]:
net.runtime_pollWait(0x7fbc65d119f0, 0x72, 0x0)
        /usr/local/go/src/pkg/runtime/netpoll.goc:146 +0x66
net.(*pollDesc).Wait(0xc20819ef40, 0x72, 0x0, 0x0)
        /usr/local/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
net.(*pollDesc).WaitRead(0xc20819ef40, 0x0, 0x0)
        /usr/local/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
net.(*netFD).accept(0xc20819eee0, 0xde51b0, 0x0, 0x7fbc65d0a3f0, 0xb)
        /usr/local/go/src/pkg/net/fd_unix.go:419 +0x343
net.(*UnixListener).AcceptUnix(0xc208158ae0, 0xc208012900, 0x0, 0x0)
        /usr/local/go/src/pkg/net/unixsock_posix.go:293 +0x73
net.(*UnixListener).Accept(0xc208158ae0, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/pkg/net/unixsock_posix.go:304 +0x4b
github.com/docker/docker/pkg/listenbuffer.(*defaultListener).Accept(0xc208158b00, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/docker/docker/pkg/listenbuffer/buffer.go:41 +0x67
net/http.(*Server).Serve(0xc2081310e0, 0x7fbc65d11aa0, 0xc208158b00, 0x0, 0x0)
        /usr/local/go/src/pkg/net/http/server.go:1698 +0x91
github.com/docker/docker/api/server.(*HttpServer).Serve(0xc208158dc0, 0x0, 0x0)
        /go/src/github.com/docker/docker/api/server/server.go:50 +0x4c
github.com/docker/docker/api/server.func·006()
        /go/src/github.com/docker/docker/api/server/server.go:1568 +0x267
created by github.com/docker/docker/api/server.ServeApi
        /go/src/github.com/docker/docker/api/server/server.go:1569 +0x3fa

goroutine 94 [runnable]:
sync.runtime_Semacquire(0xc20815e6dc)
        /usr/local/go/src/pkg/runtime/sema.goc:199 +0x30
sync.(*Mutex).Lock(0xc20815e6d8)
        /usr/local/go/src/pkg/sync/mutex.go:66 +0xd6
crypto/tls.(*Conn).sendAlert(0xc20815e580, 0x6800000265cfa000, 0x0, 0x0)
        /usr/local/go/src/pkg/crypto/tls/conn.go:684 +0x4c
crypto/tls.(*Conn).Close(0xc20815e580, 0x0, 0x0)
        /usr/local/go/src/pkg/crypto/tls/conn.go:950 +0xa9
net/http.(*persistConn).closeLocked(0xc2085a3a20)
        /usr/local/go/src/pkg/net/http/transport.go:1080 +0x55
net/http.(*persistConn).close(0xc2085a3a20)
        /usr/local/go/src/pkg/net/http/transport.go:1074 +0x70
net/http.(*persistConn).readLoop(0xc2085a3a20)
        /usr/local/go/src/pkg/net/http/transport.go:878 +0x7c0
created by net/http.(*Transport).dialConn
        /usr/local/go/src/pkg/net/http/transport.go:600 +0x93f
```

**Comment by LK4D4** [comment]
@kdomanski Thank you very much! Got this error today too.

**Comment by LK4D4** [comment]
closed by #10244
If you still experiencing this problem pls open new issue
